sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -901,12 +907,28 @@ private RunQueryResponse runQueryWithRetries(RunQueryRequest request) throws Exc
         Sleeper sleeper = Sleeper.DEFAULT;
         BackOff backoff = RUNQUERY_BACKOFF.backoff();
         while (true) {
+          HashMap<String, String> baseLabels = new HashMap<>();
+          baseLabels.put(MonitoringInfoConstants.Labels.PTRANSFORM, "TODO");
+          baseLabels.put(MonitoringInfoConstants.Labels.SERVICE, "Datastore");
+          baseLabels.put(MonitoringInfoConstants.Labels.METHOD, "runQueryWithRetries");
+          baseLabels.put(MonitoringInfoConstants.Labels.RESOURCE, "TODO");
+          baseLabels.put(
+              MonitoringInfoConstants.Labels.BIGQUERY_PROJECT_ID, request.getProjectId());
+          ServiceCallMetric serviceCallMetric =
+              new ServiceCallMetric(MonitoringInfoConstants.Urns.API_REQUEST_COUNT, baseLabels);
           try {
             RunQueryResponse response = datastore.runQuery(request);
+            serviceCallMetric.call("ok");
             rpcSuccesses.inc();
             return response;
           } catch (DatastoreException exception) {
             rpcErrors.inc();
+            GoogleJsonError.ErrorInfo errorInfo = getErrorInfo(exception);
+            if (errorInfo == null) {
+              serviceCallMetric.call(ServiceCallMetric.CANONICAL_STATUS_UNKNOWN);
+              throw exception;
+            }
+            serviceCallMetric.call(errorInfo.getReason());
```

Comment:
Lets make sure we run an integration test and exercise the error case before submitting.

I started a thread with the dev list. To see if anyone knows how to run the test.
You should be able to verify by
1. First run it normally.
2. Run it again by inducing an error (i.e trying to read from something that doesnt exist)
Add some logging to veirfy your code is called.

Reply:
Also, recently added some metric checks to unit tests that handles exceptions and retry, and the error case is setting its metric correctly.

Reply:
That's correct, please find a way to run the integration test and exercise this code end to end.
I started a thread on the dev list and added you, please follow up there to get a command to run the test.
Tweak the test (i.e. pass in a non existent entity to trigger a not found error)
Make sure your code is called with logging in both cases.
1. Normal successful case
2. Error case

Reply:
I tried the command provided in the thread but the problem that I’m having right now is that I would need a Datastore instance to try the integration test in local, do you know if there is there any guide to configure it with beam or any test datastore that I could access to do the tests?

Reply:
I don't think we can be certain that it will work unless we are able to make the system generate errors as they would be in an actual running pipeline.

Would you mind sharing the errors you get when you run the test on the thread, and following up on the issues until you are able to run the test? The dev list will know more about this than I.

I would assume the GCP apache-beam-testing project is setup to run this test by default? Do you have access to that project?

Reply:
Hi @ajamato I followed up on the dev list thread mentioning that my problem right now is that I don't have access to the GCP project to test it, I'm currently asking for access.
Meanwhile, I have a question about injecting an error, you mentioned that passing a non existing entity could be a way to do it, but passing a non existing entity would throw an exception before we attempt a read since [parsing](https://github.com/apache/beam/blob/ce406c69d2d06e5c5b659fa259253d5cc64249e7/sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/datastore/DataStoreReadWriteIT.java#L105) the query throws a ParseException beforehand. I’m trying to think on another way to cause an error but I haven’t yet thought of another possible way, most likely the DatastoreException would be thrown in cases of connection issues?

Reply:
Hi @ajamato I got the access to the GCP project and managed to run the integration tests on local adding some logs to check the conditions were passing. I’m attaching a [screenshot](https://drive.google.com/file/d/1s7_1uYNCwFLzicVM-hWFvX8-5MQjvQMZ/view?usp=sharing) of the logs of both the error case (in this case was the read operation) and ok case( in this case was the write operation). I’m also changing the PR from draft to ready to review as we are already reviewing it.

Reply:
Can you follow up on the dev list here for ideas on how to try and inject an error into the test please.

some things to consider
Are you passing an invalid query? Or just a non existent entity
Can you try referring to an entity in another project?

Reply:
Hi, I followed up in the dev list mail. I'm modifying the select query with non existent tables and fields, and the sql parser is throwing an exception before going to the Datastore reader.

Reply:
Great, glad we verified that this is working :)